### PR TITLE
Add rails versions to CI matrix

### PR DESCRIPTION
**Pull request by padde**
As [suggested](https://github.com/ruby-grape/grape-swagger-rails/pull/122#issuecomment-2344663982), here's my proposal to test against multiple Rails versions in the CI matrix. The new config will effectively run the tests against the following versions:

| Workflow | ruby | rails | grape-swagger |
| - | - | - | - |
| Tests | 2.7 | 6.1.7 | 1.6.1 |
| Tests | 3.0 | 6.1.7 | 1.6.1 |
| Tests | 3.1 | 6.1.7 | 1.6.1 |
| Tests | 3.1 | 6.1.7 | 2.0.2 |
| Tests | 3.1 | 7.2.1 | 1.6.1 |
| Tests | 3.1 | 7.2.1 | 2.0.2 |
| Tests | 3.2 | 6.1.7 | 1.6.1 |
| Tests | 3.2 | 6.1.7 | 2.0.2 |
| Tests | 3.2 | 7.2.1 | 1.6.1 |
| Tests | 3.2 | 7.2.1 | 2.0.2 |
| Tests | jruby-9.4.6 | 6.1.7 | 1.6.1 |
| Tests | jruby-9.4.6 | 6.1.7 | 2.0.2 |
| Tests | jruby-9.4.6 | 7.2.1 | 1.6.1 |
| Tests | jruby-9.4.6 | 7.2.1 | 2.0.2 |
| Tests (HEAD) | 3.2 | edge | HEAD |
| Tests (HEAD) | jruby-head | edge | HEAD |
| Tests (HEAD) | ruby-head | edge | HEAD |

### Blockers

- #124
  Fixes CI

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -6,23 +6,28 @@ jobs:
...
+        ruby: ["3.1", "3.2", "jruby-9.4.6"]
+        rails: ["6.1.7", "7.2.1"]
+        grape-swagger: ["1.6.1", "2.0.2"]
+        include:
+          - ruby: "2.7"
+            rails: "6.1.7"
+            grape-swagger: "1.6.1"
+          - ruby: "3.0"
+            rails: "6.1.7"
+            grape-swagger: "1.6.1"
```
Comment (padde) [inline .github/workflows/test.yml:18]: The versions under "include" are neither compatible with Rails >=7 nor grape-swagger >=2 so we list them separately here.

File: grape-swagger-rails.gemspec [hunk 2]
```diff
@@ -15,7 +15,11 @@ Gem::Specification.new do |spec|
   spec.license       = 'MIT'
   spec.files         = `git ls-files`.split($INPUT_RECORD_SEPARATOR)
   spec.require_paths = %w[lib]
-  spec.add_dependency 'railties', '>= 6.0.6.1'
+
+  rails_version = ENV.fetch('RAILS_VERSION', '>= 6.0.6.1')
+  rails_version = '>= 0' if rails_version == 'edge'
```
Comment (padde) [inline grape-swagger-rails.gemspec:20]: There's no way to specify a github dependency in a gemspec. Not 100% sure whether `>= 0` is the way to go here but it appears to do the trick.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -6,23 +6,28 @@ jobs:
...
-          - { ruby: "2.7", grape-swagger: "1.6.1" }
-          - { ruby: "3.0", grape-swagger: "1.6.1" }
-          - { ruby: "3.1", grape-swagger: "1.6.1" }
-          - { ruby: "3.2", grape-swagger: "1.6.1" }
-          - { ruby: "3.2", grape-swagger: "2.0.2" }
-          - { ruby: "jruby-9.4.6", grape-swagger: "1.6.1" }
-          - { ruby: "jruby-9.4.6", grape-swagger: "2.0.2" }
+        ruby: ["3.1", "3.2", "jruby-9.4.6"]
+        rails: ["6.1.7", "7.2.1"]
+        grape-swagger: ["1.6.1", "2.0.2"]
```
Comment (padde) [inline .github/workflows/test.yml:11]: The versions above this line will be tested in all possible permutations.

Comment (dblock) [reply .github/workflows/test.yml:11]: I usually find that rather wasteful as the matrix explodes with 3 different versions.

File: .github/workflows/test.yml (hunk 1 above)
Comment (dblock) [reply .github/workflows/test.yml:18]: This is why we have the other style matrix here. The full exploded combination + includes is really hard to read. Want to just list things that are compatible like they were before?

File: grape-swagger-rails.gemspec (hunk 2 above)
Comment (dblock) [reply grape-swagger-rails.gemspec:20]: Don't modify the gemspec to be dependent on ENV, that is a moving target.

Instead here say `spec.add_dependency 'railties'`, and then in Gemfile pick whichever version you want, which can be based on ENV.

**Review by dblock — CHANGES_REQUESTED** [review]
Close!

File: .github/workflows/test.yml (hunk 3 above)
Comment (padde) [reply .github/workflows/test.yml:11]: I agree that this quickly leads to larger amounts of test runs if we're not careful.

The main question here is: which combinations do you want to officially support then? Are you generally OK with the resulting combinations outlined in the PR description, keeping in mind that following the same approach will add 4 more permutations for Ruby 3.3?

Or should I instead remove some of those permutations, for example drop official support for grape-swagger 1.x after a certain Ruby version?

File: grape-swagger-rails.gemspec (hunk 2 above)
Comment (padde) [reply grape-swagger-rails.gemspec:20]: Good point. I started out with completely locking the version in here, then proceeded to override it in the Gemfile later. I think I can just simply revert the changes to the gemspec.

File: .github/workflows/test.yml (hunk 1 above)
Comment (padde) [reply .github/workflows/test.yml:18]: It probably boils down to a matter of preference, there's a tradeoff here: when adding support for a new version you have to remember to copy the right lines when adding a new Ruby version, whereas the approach I used here automatically tests agains all the supported versions of the other dependencies. I don't mind going with the previous approach though and will revert back to that.

**Comment by padde** [comment]
@dblock please check out my changes. I think the CI matrix config is fine, however some new test failures seem to be popping up that I haven't seen in the last runs, specifically with grape-swagger v2.0.2. Seems unrelated to the changes in this PR and confirmed that the old CI setup on master also fails for all grape-swagger v2.0.2 versions. There are warnings about geckodriver being incompatible with the effectively used Firefox version 130.0 (although 111.0.1 is being requested) - I'll see if I can bump those to the current versions to fix the tests.

**Comment by dblock** [comment]
Let's get https://github.com/ruby-grape/grape-swagger/issues/939 in and then we can test HEAD here with HEAD of that project too and get it all to green?

**Comment by padde** [comment]
@dblock can you please trigger a re-run on CI here? Everything should now turn green after the grape-swagger 2.1.1 release

**Comment by padde** [comment]
@dblock never mind, force push does the trick as well 😉

**Review by dblock — APPROVED** [review]

**Pull request merged**